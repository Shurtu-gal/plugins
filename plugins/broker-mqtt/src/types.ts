export interface Settings {
  mqttPort: number;
  mqttWsPort: number;
  mqttWebUiPort: number;
  mqttBrokerHost: string;
  mqttClientId: string;
  mqttPassword: string;
  mqttUsername: string;
  mqttBroker: "hivemq" | "hivemq-enterprise" | "mosquitto" | "emqx"; // @TODO: add more brokers
}
