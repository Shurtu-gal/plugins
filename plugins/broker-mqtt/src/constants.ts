import { join } from "path";

export const staticsPath = join(__dirname, "static");
export const templatesPath = join(__dirname, "templates");

const DOCKER_SERVICE_MQTT_NAME = "mqtt-broker";

export const serverPackageJsonValues = {
  dependencies: {
    "@nestjs/microservices": "^10.3.1",
    mqtt: "^5.3.5",
  },
};

export const hiveMqDockerCompose = {
  services: {
    [DOCKER_SERVICE_MQTT_NAME]: {
      image: "hivemq/hivemq4:latest",
      ports: [
        "${MQTT_PORT}:1883",
        "${MQTT_WS_PORT}:8000",
        "${MQTT_WEB_UI_PORT}:8080",
      ],
      healthcheck: {
        test: ["CMD", "echo", '""', ">", "/dev/tcp/127.0.0.1/1883"],
        interval: "5s",
        timeout: "5s",
        retries: 10,
      },
    },
  },
};

export const hivemqCommunityDockerCompose = {
  services: {
    [DOCKER_SERVICE_MQTT_NAME]: {
      image: "hivemq/hivemq-ce:latest",
      ports: ["${MQTT_PORT}:1883", "${MQTT_WS_PORT}:8000"],
      healthcheck: {
        test: ["CMD", "echo", '""', ">", "/dev/tcp/127.0.0.1/1883"],
        interval: "5s",
        timeout: "5s",
        retries: 10,
      },
    },
  },
};

export const mosquittoDockerCompose = {
  services: {
    [DOCKER_SERVICE_MQTT_NAME]: {
      image: "eclipse-mosquitto:1.6",
      ports: ["${MQTT_PORT}:1883", "${MQTT_WS_PORT}:8000"],
      environment: {
        MQTT_USER: "${MQTT_USERNAME}",
        MQTT_PASSWORD: "${MQTT_PASSWORD}",
      },
      healthcheck: {
        test: ["CMD", "echo", '""', ">", "/dev/tcp/127.0.0.1/1883"],
        interval: "5s",
        timeout: "5s",
        retries: 10,
      },
    },
  },
};

export const emqxDockerCompose = {
  services: {
    [DOCKER_SERVICE_MQTT_NAME]: {
      image: "emqx:5.4.0",
      ports: [
        "${MQTT_PORT}:1883",
        "${MQTT_WS_PORT}:8083",
        "${MQTT_WEB_UI_PORT}:18083",
      ],
      healthcheck: {
        test: ["CMD", "/opt/emqx/bin/emqx", "ctl", "status"],
        interval: "5s",
        timeout: "25s",
        retries: 5,
      },
    },
  },
};
