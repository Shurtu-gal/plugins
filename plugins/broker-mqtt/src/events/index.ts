export * from "./createServerDotEnv";
export * from "./createPackageJson";
export * from "./createAppModule";
export * from "./createDockerCompose";
