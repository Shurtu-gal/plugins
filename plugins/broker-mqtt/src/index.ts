import type {
  AmplicationPlugin,
  CreateMessageBrokerParams,
  DsgContext,
  Events,
} from "@amplication/code-gen-types";
import { EventNames } from "@amplication/code-gen-types";
import {
  afterCreateMessageBrokerClientOptionsFactory,
  afterCreateMessageBrokerNestJSModule,
  afterCreateMessageBrokerService,
  afterCreateTopicsEnum,
  beforeCreateAppModule,
  beforeCreateConnectMicroservices,
  beforeCreateDockerComposeFile,
  beforeCreateServerDotEnv,
  beforeCreateServerPackageJson,
} from "./events";
import { join } from "path";

class MQTTBrokerPlugin implements AmplicationPlugin {
  register(): Events {
    return {
      [EventNames.CreateServerDotEnv]: {
        before: beforeCreateServerDotEnv,
      },
      [EventNames.CreateServerPackageJson]: {
        before: beforeCreateServerPackageJson,
      },
      [EventNames.CreateServerDockerCompose]: {
        before: beforeCreateDockerComposeFile("PROD"),
      },
      [EventNames.CreateServerDockerComposeDev]: {
        before: beforeCreateDockerComposeFile("DEV"),
      },
      [EventNames.CreateServerAppModule]: {
        before: beforeCreateAppModule,
      },
      [EventNames.CreateMessageBroker]: {
        before: this.beforeCreateBroker,
      },
      [EventNames.CreateMessageBrokerNestJSModule]: {
        after: afterCreateMessageBrokerNestJSModule,
      },
      [EventNames.CreateMessageBrokerService]: {
        after: afterCreateMessageBrokerService,
      },
      [EventNames.CreateMessageBrokerClientOptionsFactory]: {
        after: afterCreateMessageBrokerClientOptionsFactory,
      },
      [EventNames.CreateConnectMicroservices]: {
        before: beforeCreateConnectMicroservices,
      },
      [EventNames.CreateMessageBrokerTopicsEnum]: {
        after: afterCreateTopicsEnum,
      },
    };
  }

  beforeCreateBroker(
    dsgContext: DsgContext,
    eventParams: CreateMessageBrokerParams,
  ): CreateMessageBrokerParams {
    dsgContext.serverDirectories.messageBrokerDirectory = join(
      dsgContext.serverDirectories.srcDirectory,
      "mqtt",
    );
    return eventParams;
  }
}

export default MQTTBrokerPlugin;
