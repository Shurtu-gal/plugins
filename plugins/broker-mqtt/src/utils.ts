import { VariableDictionary } from "@amplication/code-gen-types";
import { PluginInstallation } from "@amplication/code-gen-types";
import { name as PackageName } from "../package.json";
import { Settings } from "./types";
import defaultSettings from "../.amplicationrc.json";

export const getPluginSettings = (
  pluginInstallations: PluginInstallation[],
): Settings => {
  const plugin = pluginInstallations.find(
    (plugin) => plugin.npm === PackageName,
  );

  const userSettings = plugin?.settings ?? {};

  const settings: Settings = {
    ...defaultSettings.settings,
    ...userSettings,
  };

  return settings;
};

export const convertToVarDict = (
  obj: Record<string, string>,
): VariableDictionary => {
  return Object.entries(obj).map(([key, value]) => ({
    [key]: value,
  }));
};
