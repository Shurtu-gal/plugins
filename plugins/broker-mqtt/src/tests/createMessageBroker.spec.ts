import {
  BuildLogger,
  DsgContext,
  ModuleMap,
} from "@amplication/code-gen-types";
import { mock } from "jest-mock-extended";
import fg from "fast-glob";
import fs from "fs";
import {
  afterCreateMessageBrokerClientOptionsFactory,
  afterCreateMessageBrokerNestJSModule,
  afterCreateMessageBrokerService,
} from "../events";
import MQTTBrokerPlugin from "..";
import { EnumAuthProviderType } from "@amplication/code-gen-types/src/models";

describe("Testing the creation of message broker modules and files", () => {
  let context: DsgContext;
  let logger: BuildLogger;
  let plugin: MQTTBrokerPlugin;

  beforeEach(() => {
    plugin = new MQTTBrokerPlugin();
    logger = mock<BuildLogger>();
    context = mock<DsgContext>({
      pluginInstallations: [
        {
          npm: "@amplication/plugin-broker-mqtt",
        },
      ],
      serverDirectories: {
        srcDirectory: "src",
        messageBrokerDirectory: "default-message-broker",
      },
      logger: {
        info: jest.fn(),
        error: jest.fn(),
        warn: jest.fn(),
      },
      otherResources: [],
      utils: {
        importStaticModules: async (source, basePath) => {
          const directory = `${source}/`;
          const staticModules = await fg(`${directory}**/*`, {
            absolute: false,
            dot: true,
            ignore: ["**.js", "**.js.map", "**.d.ts"],
          });

          const modules = await Promise.all(
            staticModules.sort().map(async (module) => {
              return {
                path: module.replace(directory, basePath ? basePath + "/" : ""),
                code: await fs.promises.readFile(module, "utf-8"),
              };
            }),
          );

          const moduleMap: ModuleMap = new ModuleMap(logger);

          for await (const module of modules) {
            await moduleMap.set(module);
          }

          return moduleMap;
        },
      },
    });
  });

  it("all modules and files should be created with default values", async () => {
    const modules = new ModuleMap(logger);
    plugin.beforeCreateBroker(context, modules);
    const brokerModules = await afterCreateMessageBrokerService(context);
    const generateClientOptionsModule =
      await afterCreateMessageBrokerClientOptionsFactory(context);
    const nestjsModule = await afterCreateMessageBrokerNestJSModule(context);

    await modules.mergeMany([
      brokerModules,
      generateClientOptionsModule,
      nestjsModule,
    ]);

    expect(modules.modules()).toMatchSnapshot();
  });

  it("all modules and files should be created with custom values", async () => {
    const modules = new ModuleMap(logger);
    context.otherResources = [
      {
        buildId: "custom-mqtt-id",
        pluginInstallations: [],
        resourceType: "MessageBroker",
        resourceInfo: {
          name: "custom-mqtt",
          codeGeneratorVersionOptions: {
            codeGeneratorStrategy: undefined,
            codeGeneratorVersion: undefined,
          },
          description: "Custom MQTT broker",
          id: "custom-mqtt-id",
          settings: {
            adminUISettings: {
              adminUIPath: "/admin",
              generateAdminUI: true,
            },
            authProvider: EnumAuthProviderType.Jwt,
            serverSettings: {
              serverPath: "/server",
              generateGraphQL: true,
              generateRestApi: true,
            },
          },
          url: "mqtt://localhost:1883",
          version: "1.0.0",
        },
      },
    ];
    plugin.beforeCreateBroker(context, modules);
    const brokerModules = await afterCreateMessageBrokerService(context);
    const generateClientOptionsModule =
      await afterCreateMessageBrokerClientOptionsFactory(context);
    const nestjsModule = await afterCreateMessageBrokerNestJSModule(context);

    await modules.mergeMany([
      brokerModules,
      generateClientOptionsModule,
      nestjsModule,
    ]);

    expect(modules.modules()).toMatchSnapshot();
  });

  it("all modules and files should be created with sparkplug enabled", async () => {
    context.pluginInstallations[0].settings = {
      sparkplugConfig: {
        enabled: true,
      },
    };
    const modules = new ModuleMap(logger);
    plugin.beforeCreateBroker(context, modules);
    const brokerModules = await afterCreateMessageBrokerService(context);
    const generateClientOptionsModule =
      await afterCreateMessageBrokerClientOptionsFactory(context);
    const nestjsModule = await afterCreateMessageBrokerNestJSModule(context);

    await modules.mergeMany([
      brokerModules,
      generateClientOptionsModule,
      nestjsModule,
    ]);

    expect(modules.modules()).toMatchSnapshot();
  });
});
