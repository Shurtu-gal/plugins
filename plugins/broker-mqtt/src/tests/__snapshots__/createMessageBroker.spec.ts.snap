// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing the creation of message broker modules and files all modules and files should be created with custom values 1`] = `
[
  {
    "code": "import { ClientProxy } from "@nestjs/microservices";
import { MqttMessage } from "./types";
import { Inject, OnModuleInit } from "@nestjs/common";
import { CustomMqttTopics } from "./topics";

export class MqttProducerService implements OnModuleInit {
  constructor(
    @Inject("MQTT_CLIENT") private readonly mqttClient: ClientProxy,
  ) {}

  async publish(topic: CustomMqttTopics, message: MqttMessage): Promise<void> {
    return await new Promise((resolve, reject) => {
      this.mqttClient.emit(topic, message).subscribe({
        error(err) {
          reject(err);
        },
        next() {
          resolve();
        },
      });
    });
  }

  async onModuleInit() {
    await this.mqttClient.connect();
  }
}
",
    "path": "src/mqtt/mqtt.producer.service.ts",
  },
  {
    "code": "export type MqttMessage = string | Record<string, unknown> | null;  

export interface MqttMessageHeaders {
  [key: string]: Buffer | string | number | boolean | null;
}",
    "path": "src/mqtt/types.ts",
  },
  {
    "code": "import { ConfigService } from "@nestjs/config";
import { Transport, MqttOptions } from "@nestjs/microservices";

export const generateClientOptions = (
  configService: ConfigService
): MqttOptions => {
  const mqttBrokerHost = configService.get<string>("MQTT_BROKER_HOST");
  const mqttBrokerPort = configService.get<number>("MQTT_PORT");
  const clientId = configService.get<string>("MQTT_CLIENT_ID");
  const username = configService.get<string>("MQTT_USERNAME");
  const password = configService.get<string>("MQTT_PASSWORD");
  
  if (!mqttBrokerHost || !mqttBrokerPort || !clientId || !username || !password) {
    throw new Error(
      "MQTT_BROKER_HOST, MQTT_PORT, MQTT_CLIENT_ID, MQTT_USERNAME, MQTT_PASSWORD must be defined"
    );
  }

  return {
    transport: Transport.MQTT,
    options: {
      host: mqttBrokerHost,
      port: mqttBrokerPort,
      clientId: clientId,
      username,
      password,
      connectTimeout: 10000,
    },
  };
}",
    "path": "src/mqtt/generateClientOptions.ts",
  },
  {
    "code": "import { Global, Module } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { ClientProxyFactory } from "@nestjs/microservices";
import { generateClientOptions } from "./generateClientOptions";
import { MqttProducerService } from "./mqtt.producer.service";
import { MqttController } from "./mqtt.controller";

@Global()
@Module({
  imports: [],
  providers: [
    {
      provide: "MQTT_CLIENT",
      useFactory: (configService: ConfigService) => {
        return ClientProxyFactory.create(
          generateClientOptions(configService),
        );
      },
      inject: [ConfigService],
    },
    MqttProducerService
  ],
  controllers: [MqttController],
  exports: [MqttProducerService],
})

export class MqttModule {}",
    "path": "src/mqtt/mqtt.module.ts",
  },
]
`;

exports[`Testing the creation of message broker modules and files all modules and files should be created with default values 1`] = `
[
  {
    "code": "import { ClientProxy } from "@nestjs/microservices";
import { MqttMessage } from "./types";
import { Inject, OnModuleInit } from "@nestjs/common";
import { MqttTopics } from "./topics";

export class MqttProducerService implements OnModuleInit {
  constructor(
    @Inject("MQTT_CLIENT") private readonly mqttClient: ClientProxy,
  ) {}

  async publish(topic: MqttTopics, message: MqttMessage): Promise<void> {
    return await new Promise((resolve, reject) => {
      this.mqttClient.emit(topic, message).subscribe({
        error(err) {
          reject(err);
        },
        next() {
          resolve();
        },
      });
    });
  }

  async onModuleInit() {
    await this.mqttClient.connect();
  }
}
",
    "path": "src/mqtt/mqtt.producer.service.ts",
  },
  {
    "code": "export type MqttMessage = string | Record<string, unknown> | null;  

export interface MqttMessageHeaders {
  [key: string]: Buffer | string | number | boolean | null;
}",
    "path": "src/mqtt/types.ts",
  },
  {
    "code": "import { ConfigService } from "@nestjs/config";
import { Transport, MqttOptions } from "@nestjs/microservices";

export const generateClientOptions = (
  configService: ConfigService
): MqttOptions => {
  const mqttBrokerHost = configService.get<string>("MQTT_BROKER_HOST");
  const mqttBrokerPort = configService.get<number>("MQTT_PORT");
  const clientId = configService.get<string>("MQTT_CLIENT_ID");
  const username = configService.get<string>("MQTT_USERNAME");
  const password = configService.get<string>("MQTT_PASSWORD");
  
  if (!mqttBrokerHost || !mqttBrokerPort || !clientId || !username || !password) {
    throw new Error(
      "MQTT_BROKER_HOST, MQTT_PORT, MQTT_CLIENT_ID, MQTT_USERNAME, MQTT_PASSWORD must be defined"
    );
  }

  return {
    transport: Transport.MQTT,
    options: {
      host: mqttBrokerHost,
      port: mqttBrokerPort,
      clientId: clientId,
      username,
      password,
      connectTimeout: 10000,
    },
  };
}",
    "path": "src/mqtt/generateClientOptions.ts",
  },
  {
    "code": "import { Global, Module } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { ClientProxyFactory } from "@nestjs/microservices";
import { generateClientOptions } from "./generateClientOptions";
import { MqttProducerService } from "./mqtt.producer.service";
import { MqttController } from "./mqtt.controller";

@Global()
@Module({
  imports: [],
  providers: [
    {
      provide: "MQTT_CLIENT",
      useFactory: (configService: ConfigService) => {
        return ClientProxyFactory.create(
          generateClientOptions(configService),
        );
      },
      inject: [ConfigService],
    },
    MqttProducerService
  ],
  controllers: [MqttController],
  exports: [MqttProducerService],
})

export class MqttModule {}",
    "path": "src/mqtt/mqtt.module.ts",
  },
]
`;

exports[`Testing the creation of message broker modules and files all modules and files should be created with sparkplug enabled 1`] = `
[
  {
    "code": "import { ClientProxy } from "@nestjs/microservices";
import { MqttMessage } from "./types";
import { Inject, OnModuleInit } from "@nestjs/common";
import { MqttTopics } from "./topics";

export class MqttProducerService implements OnModuleInit {
  constructor(
    @Inject("MQTT_CLIENT") private readonly mqttClient: ClientProxy,
  ) {}

  async publish(topic: MqttTopics, message: MqttMessage): Promise<void> {
    return await new Promise((resolve, reject) => {
      this.mqttClient.emit(topic, message).subscribe({
        error(err) {
          reject(err);
        },
        next() {
          resolve();
        },
      });
    });
  }

  async onModuleInit() {
    await this.mqttClient.connect();
  }
}
",
    "path": "src/mqtt/mqtt.producer.service.ts",
  },
  {
    "code": "export type MqttMessage = string | Record<string, unknown> | null;  

export interface MqttMessageHeaders {
  [key: string]: Buffer | string | number | boolean | null;
}",
    "path": "src/mqtt/types.ts",
  },
  {
    "code": "import { ConfigService } from "@nestjs/config";
import { Transport, MqttOptions } from "@nestjs/microservices";

export const generateClientOptions = (
  configService: ConfigService
): MqttOptions => {
  const mqttBrokerHost = configService.get<string>("MQTT_BROKER_HOST");
  const mqttBrokerPort = configService.get<number>("MQTT_PORT");
  const clientId = configService.get<string>("MQTT_CLIENT_ID");
  const username = configService.get<string>("MQTT_USERNAME");
  const password = configService.get<string>("MQTT_PASSWORD");
  
  if (!mqttBrokerHost || !mqttBrokerPort || !clientId || !username || !password) {
    throw new Error(
      "MQTT_BROKER_HOST, MQTT_PORT, MQTT_CLIENT_ID, MQTT_USERNAME, MQTT_PASSWORD must be defined"
    );
  }

  return {
    transport: Transport.MQTT,
    options: {
      host: mqttBrokerHost,
      port: mqttBrokerPort,
      clientId: clientId,
      username,
      password,
      connectTimeout: 10000,
    },
  };
}",
    "path": "src/mqtt/generateClientOptions.ts",
  },
  {
    "code": "import { Global, Module } from "@nestjs/common";
import { ConfigService } from "@nestjs/config";
import { ClientProxyFactory } from "@nestjs/microservices";
import { generateClientOptions } from "./generateClientOptions";
import { MqttProducerService } from "./mqtt.producer.service";
import { MqttController } from "./mqtt.controller";
import { SparkplugService } from './sparkplug.service';

@Global()
@Module({
  imports: [],
  providers: [
    {
      provide: "MQTT_CLIENT",
      useFactory: (configService: ConfigService) => {
        return ClientProxyFactory.create(
          generateClientOptions(configService),
        );
      },
      inject: [ConfigService],
    },
    MqttProducerService,
    SparkplugService
  ],
  controllers: [MqttController],
  exports: [MqttProducerService, SparkplugService],
})

export class MqttModule {}",
    "path": "src/mqtt/mqtt.module.ts",
  },
  {
    "code": "import { Injectable, Logger } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
var SparkPlug = require('sparkplug-client');

@Injectable()
export class SparkplugServiceBase {
  public client: ReturnType<typeof SparkPlug.newClient>;
  private readonly logger = new Logger(SparkplugServiceBase.name);

  constructor(protected readonly configService: ConfigService) {
    const serverUrl = \`mqtt://\${this.configService.get('MQTT_BROKER_HOST')}:\${this.configService.get('MQTT_PORT')}\`;

    const config = {
      clientId : this.configService.get('MQTT_SPARKPLUG_CLIENT_ID')!,
      username : this.configService.get('MQTT_USERNAME')!,
      password : this.configService.get('MQTT_PASSWORD')!,
      groupId : this.configService.get('MQTT_SPARKPLUG_GROUP_ID')!,
      edgeNode: this.configService.get('MQTT_SPARKPLUG_EDGE_NODE')!,
      serverUrl,
    };

    this.client = SparkPlug.newClient(config);

    this.client.on('connect', () => {
      this.logger.log('Connected to MQTT broker');
    });

    this.client.on('error', (err: Error) => {
      this.logger.error(\`Error: \${err}\`);
    });
  }
}
",
    "path": "src/mqtt/base/sparkplug.service.base.ts",
  },
  {
    "code": "import { Injectable } from '@nestjs/common';
import sparkPlug from 'sparkplug-client';
import { SparkplugServiceBase } from './base/sparkplug.service.base';
import { ConfigService } from '@nestjs/config';

@Injectable()
export class SparkplugService extends SparkplugServiceBase {
  constructor(protected readonly configService: ConfigService) {
    super(configService);
  }
}
",
    "path": "src/mqtt/sparkplug.service.ts",
  },
]
`;
